// game/index.js
// Generated by CoffeeScript 1.6.1
var AI, BattleRound, Card, CardPile, GAME_STATE, Game, MOVE_TYPE, Move, Player, PlayerCardPile, PlayerInventoryCardPile, Table, UnusedCardPile, UsedCardPile,
  __hasProp = {}.hasOwnProperty,
  __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

MOVE_TYPE = {
  STAY: 1,
  HIT: 2,
  SPLIT: 3,
  BUST: 4,
  QUIT: 5
};

GAME_STATE = {
  NEW_GAME: 1,
  ROUND_START: 2,
  DEFENDER_MOVING: 3,
  ATTACKER_MOVING: 4,
  ROUND_END: 5,
  GAME_OVER: 6
};

Game = (function() {

  function Game() {
    this.table = new Table();
    this.players = [new Player(), new Player()];
    this.thisPlayer = 0;
    this.lastMoveId = 0;
    this.currentRound = null;
    this.state = GAME_STATE.NEW_GAME;
    this.moves = [];
    this.dealNextHand();
  }

  Game.prototype.getSaveData = function() {
    return {
      table: this.table.getSaveData(),
      state: this.state,
      currentRound: this.currentRound.getSaveData(),
      players: [this.players[0].getSaveData(), this.players[1].getSaveData()]
    };
  };

  Game.prototype.loadData = function(data) {
    var attackerHand, i, player, _i, _len, _ref;
    this.table.loadData(data.table);
    this.state = data.state;
    this.currentRound.loadData(data.currentRound);
    _ref = this.players;
    for (i = _i = 0, _len = _ref.length; _i < _len; i = ++_i) {
      player = _ref[i];
      player.loadData(data.players[i]);
    }
    if (this.state === GAME_STATE.ATTACKER_MOVING) {
      attackerHand = this.table.playerHands[this.currentRound.playerTurn];
      return attackerHand.reveal();
    }
  };

  Game.prototype.dealNextHand = function() {
    var attackerHand, defenderHand, startingPlayer;
    if (this.state === GAME_STATE.GAME_OVER) {
      return;
    }
    this.state = GAME_STATE.ROUND_START;
    if (this.currentRound != null) {
      this.clearHand();
      startingPlayer = this.currentRound.playerTurn;
    } else {
      startingPlayer = 0;
    }
    defenderHand = this.table.playerHands[startingPlayer];
    attackerHand = this.table.playerHands[Number(!startingPlayer)];
    attackerHand.isRevealingAll = false;
    defenderHand.push(this.table.unusedPile.takeNextCard());
    attackerHand.push(this.table.unusedPile.takeNextCard());
    defenderHand.push(this.table.unusedPile.takeNextCard());
    attackerHand.push(this.table.unusedPile.takeNextCard());
    this.currentRound = new BattleRound(startingPlayer);
    return this.state = GAME_STATE.DEFENDER_MOVING;
  };

  Game.prototype.makeMove = function(move) {
    var attackerHand, card, playerHand;
    if (this.state === GAME_STATE.GAME_OVER) {
      return;
    }
    this.moves.push(new Move(this.currentRound.playerTurn, move));
    this.currentRound.makeMove(move);
    switch (move) {
      case MOVE_TYPE.HIT:
        card = this.table.unusedPile.takeNextCard();
        playerHand = this.table.playerHands[this.currentRound.playerTurn];
        playerHand.push(card);
        if (playerHand.value() > 21) {
          this.makeMove(MOVE_TYPE.BUST);
        }
        break;
      case MOVE_TYPE.STAY:
      case MOVE_TYPE.BUST:
        if (this.currentRound.isOver) {
          console.log('Battle Round Is Over');
          this.evaluateResult();
          this.dealNextHand();
        } else {
          this.state = GAME_STATE.ATTACKER_MOVING;
          attackerHand = this.table.playerHands[this.currentRound.playerTurn];
          attackerHand.reveal();
        }
        break;
      default:
        null;
    }
    return console.log('moves so far', this.moves);
  };

  Game.prototype.evaluateResult = function() {
    var hpLost, player1, player1HandValue, player2, player2HandValue;
    player1 = this.players[0];
    player1HandValue = this.table.playerHands[0].value();
    if (player1HandValue > 21) {
      player1HandValue = 0;
    }
    player2 = this.players[1];
    player2HandValue = this.table.playerHands[1].value();
    if (player2HandValue > 21) {
      player2HandValue = 0;
    }
    if (player1HandValue > player2HandValue) {
      hpLost = player1HandValue - player2HandValue;
      if (hpLost > 11) {
        hpLost = 11;
      }
      player2.takeDamage(hpLost);
    } else if (player2HandValue > player1HandValue) {
      hpLost = player2HandValue - player1HandValue;
      if (hpLost > 11) {
        hpLost = 11;
      }
      player1.takeDamage(hpLost);
    } else {
      console.log('tie round, no damage');
    }
    if (player1.hp === 0 || player2.hp === 0) {
      return this.state = GAME_STATE.GAME_OVER;
    }
  };

  Game.prototype.clearHand = function() {
    var card, cards, hand, _i, _len, _ref, _results;
    _ref = this.table.playerHands;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      hand = _ref[_i];
      cards = hand.takeAllCards();
      _results.push((function() {
        var _j, _len1, _results1;
        _results1 = [];
        for (_j = 0, _len1 = cards.length; _j < _len1; _j++) {
          card = cards[_j];
          _results1.push(this.table.usedPile.push(card));
        }
        return _results1;
      }).call(this));
    }
    return _results;
  };

  return Game;

})();

/* --------------------------------------------
     Begin Card.coffee
--------------------------------------------
*/


Card = (function() {

  function Card(value, suit) {
    this.value = value;
    this.suit = suit;
    this.currentPile = null;
    this.isFlipped = false;
  }

  Card.prototype.getSaveData = function() {
    return {
      value: this.value,
      suit: this.suit
    };
  };

  Card.prototype.loadData = function(data) {
    this.value = data.value;
    return this.suit = data.suit;
  };

  return Card;

})();

/* --------------------------------------------
     Begin CardPile.coffee
--------------------------------------------
*/


CardPile = (function() {

  function CardPile() {
    this.cards = [];
    this.positionTop = 0;
    this.positionLeft = 0;
  }

  CardPile.prototype.cardPosition = function(card) {
    return {
      top: this.positionTop,
      left: this.positionLeft
    };
  };

  CardPile.prototype.shuffle = function() {
    /*
    Fisher Yates Shuffle algorithm for arrays
    */

    var i, j, tempi, tempj, _results;
    i = this.cards.length;
    if (i === 0) {
      return false;
    }
    _results = [];
    while (--i) {
      j = Math.floor(Math.random() * (i + 1));
      tempi = this.cards[i];
      tempj = this.cards[j];
      this.cards[i] = tempj;
      _results.push(this.cards[j] = tempi);
    }
    return _results;
  };

  CardPile.prototype.push = function(card) {
    card.currentPile = this;
    return this.cards.push(card);
  };

  CardPile.prototype.takeNextCard = function() {
    return this.cards.shift();
  };

  CardPile.prototype.takeAllCards = function() {
    return this.cards.splice(0);
  };

  CardPile.prototype.getSaveData = function() {
    var card, cards, _i, _len, _ref;
    cards = [];
    _ref = this.cards;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      card = _ref[_i];
      cards.push(card.getSaveData());
    }
    return {
      cards: cards
    };
  };

  CardPile.prototype.loadData = function(data) {
    var card, _i, _len, _ref, _results;
    this.cards.length = 0;
    _ref = data.cards;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      card = _ref[_i];
      card = new Card(card.value, card.suit);
      _results.push(this.push(card));
    }
    return _results;
  };

  return CardPile;

})();

UnusedCardPile = (function(_super) {

  __extends(UnusedCardPile, _super);

  function UnusedCardPile() {
    UnusedCardPile.__super__.constructor.call(this);
    this.positionTop = 160 + "px";
    this.positionLeft = 0;
  }

  return UnusedCardPile;

})(CardPile);

UsedCardPile = (function(_super) {

  __extends(UsedCardPile, _super);

  function UsedCardPile() {
    UsedCardPile.__super__.constructor.call(this);
    this.positionTop = 160 + "px";
    this.positionLeft = 270 + "px";
  }

  UsedCardPile.prototype.push = function(card) {
    UsedCardPile.__super__.push.call(this, card);
    return card.isFlipped = false;
  };

  return UsedCardPile;

})(CardPile);

PlayerCardPile = (function(_super) {

  __extends(PlayerCardPile, _super);

  function PlayerCardPile(positionTop, positionLeft) {
    PlayerCardPile.__super__.constructor.call(this);
    this.positionTop = positionTop;
    this.positionLeft = positionLeft;
    this.isRevealingAll = true;
  }

  PlayerCardPile.prototype.push = function(card) {
    PlayerCardPile.__super__.push.call(this, card);
    if (this.isRevealingAll || this.cards.length === 1) {
      return card.isFlipped = true;
    }
  };

  PlayerCardPile.prototype.reveal = function() {
    var card, _i, _len, _ref, _results;
    this.isRevealingAll = true;
    _ref = this.cards;
    _results = [];
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      card = _ref[_i];
      _results.push(card.isFlipped = true);
    }
    return _results;
  };

  PlayerCardPile.prototype.value = function() {
    var aceCount, card, handValue, i, _i, _j, _len, _ref;
    if (!this.cards.length) {
      return 0;
    }
    if (!this.isRevealingAll) {
      return "?";
    }
    handValue = 0;
    aceCount = 0;
    _ref = this.cards;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      card = _ref[_i];
      switch (card.value) {
        case "A":
          aceCount++;
          break;
        case "J":
        case "Q":
        case "K":
          handValue += 10;
          break;
        default:
          handValue += card.value;
      }
    }
    if (aceCount != null) {
      for (i = _j = 0; 0 <= aceCount ? _j < aceCount : _j > aceCount; i = 0 <= aceCount ? ++_j : --_j) {
        if (handValue <= 10) {
          handValue += 11;
        } else {
          handValue += 1;
        }
      }
    }
    return handValue;
  };

  PlayerCardPile.prototype.cardPosition = function(card) {
    var i, pileCard, _i, _len, _ref;
    _ref = this.cards;
    for (i = _i = 0, _len = _ref.length; _i < _len; i = ++_i) {
      pileCard = _ref[i];
      if (pileCard === card) {
        return {
          "top": this.positionTop + "px",
          "left": this.positionLeft + (i * 30 - this.cards.length * 15) + "px",
          "z-index": i
        };
      }
    }
  };

  return PlayerCardPile;

})(CardPile);

PlayerInventoryCardPile = (function(_super) {

  __extends(PlayerInventoryCardPile, _super);

  function PlayerInventoryCardPile() {
    return PlayerInventoryCardPile.__super__.constructor.apply(this, arguments);
  }

  return PlayerInventoryCardPile;

})(CardPile);

/* --------------------------------------------
     Begin Player.coffee
--------------------------------------------
*/


Player = (function() {

  function Player() {
    this.hp = 20;
    this.maxHp = 20;
  }

  Player.prototype.getHp = function() {
    return this.hp;
  };

  Player.prototype.takeDamage = function(dmg) {
    console.log('Player Taking Damage', dmg);
    this.hp -= dmg;
    if (this.hp < 0) {
      return this.hp = 0;
    }
  };

  Player.prototype.getSaveData = function() {
    return {
      hp: this.hp,
      maxHp: this.maxHp
    };
  };

  Player.prototype.loadData = function(data) {
    this.hp = data.hp;
    return this.maxHp = data.maxHp;
  };

  return Player;

})();

/* --------------------------------------------
     Begin BattleRound.coffee
--------------------------------------------
*/


BattleRound = (function() {

  function BattleRound(playerTurn) {
    this.isOver = false;
    this.playerTurn = playerTurn;
    this.playerCompleteStatuses = [false, false];
    this.movesMade = [];
  }

  BattleRound.prototype.makeMove = function(move) {
    this.movesMade.push(move);
    if (move === MOVE_TYPE.STAY || move === MOVE_TYPE.BUST) {
      this.playerCompleteStatuses[this.playerTurn] = true;
      if (this.playerCompleteStatuses[0] && this.playerCompleteStatuses[1]) {
        return this.isOver = true;
      } else {
        return this.playerTurn = Number(!this.playerTurn);
      }
    }
  };

  BattleRound.prototype.getSaveData = function() {
    return {
      isOver: this.isOver,
      playerTurn: this.playerTurn,
      playerCompleteStatuses: this.playerCompleteStatuses,
      movesMade: this.movesMade
    };
  };

  BattleRound.prototype.loadData = function(data) {
    this.isOver = data.isOver;
    this.playerTurn = data.playerTurn;
    this.playerCompleteStatuses = data.playerCompleteStatuses;
    return this.movesMade = data.movesMade;
  };

  return BattleRound;

})();

/* --------------------------------------------
     Begin Move.coffee
--------------------------------------------
*/


Move = (function() {

  function Move(player, move_type) {
    this.player = player;
    this.move_type = move_type;
    this.date = Date();
  }

  return Move;

})();

/* --------------------------------------------
     Begin Table.coffee
--------------------------------------------
*/


Table = (function() {

  function Table() {
    var card, suit, value, _i, _j, _len, _len1, _ref, _ref1;
    this.suits = ["spades", "clubs", "diamonds", "hearts"];
    this.values = [2, 3, 4, 5, 6, 7, 8, 9, 10].concat(["J", "Q", "K", "A"]);
    this.allCards = [];
    this.unusedPile = new UnusedCardPile();
    this.usedPile = new UsedCardPile();
    this.playerHands = [new PlayerCardPile(300, 100), new PlayerCardPile(40, 100)];
    this.playerInventories = [new PlayerInventoryCardPile(), new PlayerInventoryCardPile()];
    _ref = this.suits;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      suit = _ref[_i];
      _ref1 = this.values;
      for (_j = 0, _len1 = _ref1.length; _j < _len1; _j++) {
        value = _ref1[_j];
        card = new Card(value, suit);
        this.allCards.push(card);
        this.unusedPile.push(card);
      }
    }
    this.unusedPile.shuffle();
  }

  Table.prototype.getSaveData = function() {
    return {
      unusedPile: this.unusedPile.getSaveData(),
      usedPile: this.usedPile.getSaveData(),
      playerHands: [this.playerHands[0].getSaveData(), this.playerHands[1].getSaveData()]
    };
  };

  Table.prototype.loadData = function(data) {
    var card, i, playerHand, _i, _j, _k, _len, _len1, _len2, _ref, _ref1, _ref2, _results;
    this.allCards.length = 0;
    this.unusedPile.loadData(data.unusedPile);
    _ref = this.unusedPile.cards;
    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
      card = _ref[_i];
      this.allCards.push(card);
    }
    this.usedPile.loadData(data.usedPile);
    _ref1 = this.usedPile;
    for (_j = 0, _len1 = _ref1.length; _j < _len1; _j++) {
      card = _ref1[_j];
      this.allCards.push(card);
    }
    _ref2 = this.playerHands;
    _results = [];
    for (i = _k = 0, _len2 = _ref2.length; _k < _len2; i = ++_k) {
      playerHand = _ref2[i];
      playerHand.loadData(data.playerHands[i]);
      _results.push((function() {
        var _l, _len3, _ref3, _results1;
        _ref3 = playerHand.cards;
        _results1 = [];
        for (_l = 0, _len3 = _ref3.length; _l < _len3; _l++) {
          card = _ref3[_l];
          _results1.push(this.allCards.push(card));
        }
        return _results1;
      }).call(this));
    }
    return _results;
  };

  return Table;

})();

/* --------------------------------------------
     Begin AI.coffee
--------------------------------------------
*/


AI = (function() {

  function AI() {}

  AI.prototype.decide = function(aiCardPile, playerCardPile) {
    if (playerCardPile.value() > 21) {
      return MOVE_TYPE.STAY;
    }
    if (aiCardPile.value() < 17) {
      return MOVE_TYPE.HIT;
    } else {
      return MOVE_TYPE.STAY;
    }
  };

  return AI;

})();

/* --------------------------------------------
     Begin index.coffee
--------------------------------------------
*/


module.exports.Game = Game;

module.exports.AI = AI;
